src/content/blog/summer-of-codegen.mdx

```diff
@@ -0,0 +1,440 @@
+  <a class="underline hover:no-underline" href="https://${username}-${tempValName}.web.val.run">${username}/${tempValName}</a>.
+</div>`);
+```
+
+That's all it takes to generate a val:
+
+<figure style={{ maxWidth: "600px", border: '1px solid var(--border)' }}>
+  <Image src={HelloworldgenImage} alt="" />
+</figure>
+
+And when opening the link at the bottom:
+
+<figure style={{ maxWidth: "600px", border: '1px solid var(--border)' }}>
+  <Image src={HelloworldappImage} alt="" />
+</figure>
+
+We can even embrace the "HTML energy", and hold ctrl/cmd while clicking "Go", to get multiple tabs generating vals:
+
+<figure style={{ maxWidth: "600px" }}>
+  <video controls loop>
+    <source src="/video/htmlenergy.mp4" />
+  </video>
+  <figcaption>
+  </figcaption>
+</figure>
+
+# Advanced prototype
+
+Over the course of several weeks, I ended up adding lots of features to this prototype:
+
+* Side panel with code and preview. The code even streams in by adding a `<script>` tag for each token, that calls a function adding the token to the code editor (you end up with very cursed HTML, with a gazillion script tags).
+* CodeMirror for syntax highlighting.
+* Follow-up prompts to iterate. This generates a new val for every iteration, so you can easily go back, or generate multiple vals by holding ctrl/cmd.
+* Loading an existing val and iterating on that.
+* Editing code manually.
+* Switching between models (Claude, OpenAI) and context window size.
+
+<figure style={{ maxWidth: "600px" }}>
+  <video controls loop>
+    <source src="/video/valle-hackernews.mp4" />
+  </video>
+  <figcaption>
+  </figcaption>
+</figure>
+
+I called this prototype VALL-E, and you can [fork it yourself](https://www.val.town/v/janpaul123/valle). Let's look at the details of what we ran into.
+
+### Blob storage
+
+The whole benefit of codegen on Val Town is that we can instantly deploy backend code. But we have some particular APIs that LLMs are not familiar with.
+
+One of these is our [blob storage](https://docs.val.town/std/blob/). We teach LLMs about blob storage by adding some example vals in its context window, as well as the full markdown of the blob storage docs page. Still, it would make minor mistakes such as using common keys that can clash with other vals, so we explicitly told it how to avoid that, by adding to its prompt:
+
+```
+Use the current val name for the storage key, which you can get
+using the "pomdtr/extractValInfo" val, which you can call like this:
+\`\`\`ts
+const {author, name} = extractValInfo(import.meta.url);
+\`\`\`
+```
+
+### SQLite
+
+Each val has a [SQLite](https://docs.val.town/std/sqlite/) database, however LLMs had trouble with its API, since it uses arrays (instead of objects) as inputs and outputs. LLMs would consistently tripped up on this, so we told them to avoid SQLite for now. Instead of more prompting, we will soon write a wrapper that makes the SQLite API look more like what LLMs expect — this generally works better.
```

Comment:
I might expand this section and talk about how turso has an unstandard sqlite interface and it makes the llm trip up. Our strategy is to build a shim over it that is more familiar to the LLM. I think that's a core insight – bend to the LLM instead of prompt around it